# wf-paired-end-illumina-assembly: Citations

## [nf-core](https://pubmed.ncbi.nlm.nih.gov/32055031/)

> Ewels PA, Peltzer A, Fillinger S, Patel H, Alneberg J, Wilm A, Garcia MU, Di Tommaso P, Nahnsen S. The nf-core framework for community-curated bioinformatics pipelines. Nat Biotechnol. 2020 Mar;38(3):276-278. doi: 10.1038/s41587-020-0439-x. PubMed PMID: 32055031.

## [Nextflow](https://pubmed.ncbi.nlm.nih.gov/28398311/)

> Di Tommaso P, Chatzou M, Floden EW, Barja PP, Palumbo E, Notredame C. Nextflow enables reproducible computational workflows. Nat Biotechnol. 2017 Apr 11;35(4):316-319. doi: 10.1038/nbt.3820. PubMed PMID: 28398311.

## Pipeline tools

- [AWK](https://www.thriftbooks.com/w/the-awk-programming-language_brian-w-kernighan_alfred-v-aho/254416/?resultid=8ae6bc1c-db73-4eb0-b297-9d28a08ee38f#edition=2350735&idiq=2920572)

  > Aho AV, Kernighan BW, Weinberger PJ. The AWK programming language. 224 pp. Pearson. ISBN-13: 978-0201079814

- [Bakta](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8743544/)

  > Schwengers O, Jelonek L, Dieckmann MA, Beyvers S, Blom J, Goesmann A. Bakta: rapid and standardized annotation of bacterial genomes via alignment-free sequence identification. Microb Genom. Nov 2021;7(11):000685. doi: 10.1099/mgen.0.000685

- [BEDTools](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4213956/)

  > Quinlan AR. BEDTools: The Swiss-Army Tool for Genome Feature Analysis. Curr Protoc Bioinformatics. Sep 8 2014;47:11.12.1-34. doi:10.1002/0471250953.bi1112s47

- [BLAST+](https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-10-421)

  > Camacho C, Coulouris G, Avagyan V, et al. BLAST+: architecture and applications. BMC Bioinformatics. 2009/12/15 2009;10(1):421. doi:10.1186/1471-2105-10-421

- [BUSCO](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8476166/)

  > Manni M, Berkeley MR, Seppey M, Simão FA, Zdobnov EM. BUSCO Update: Novel and Streamlined Workflows along with Broader and Deeper Phylogenetic Coverage for Scoring of Eukaryotic, Prokaryotic, and Viral Genomes. Mol Biol Evol. Sep 27 2021;38(10):4647-4654. doi:10.1093/molbev/msab199

- [BWA-MEM](https://arxiv.org/abs/1303.3997)

  > Li H. Aligning sequence reads, clone sequences and assembly contigs with BWA-MEM. May 26 2013; arXiv:1303.3997. doi:10.48550/arXiv.1303.3997

- [BioPython](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2682512/)

  > Cock PJ, Antao T, Chang JT, et al. Biopython: freely available Python tools for computational molecular biology and bioinformatics. Bioinformatics. Jun 1 2009;25(11):1422-3. doi:10.1093/bioinformatics/btp163

- [CAT](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6805573/)

  > von Meijenfeldt FAB, Arkhipova K, Cambuy DD, Coutinho FH, Dutilh BE. Robust taxonomic classification of uncharted microbial sequences and bins with CAT and BAT. Genome Biol. Oct 22 2019;20(1):217. doi: 10.1186/s13059-019-1817-x

- [CheckM2](https://pubmed.ncbi.nlm.nih.gov/37500759/)

  > Chklovski A, Parks DH, Woodcroft BJ, Tyson GW. CheckM2: a rapid, scalable and accurate tool for assessing microbial genome quality using machine learning. Nat Methods. Aug 2023;20(8):1203-1212. doi: 10.1038/s41592-023-01940-w

- [Fastp](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6129281/)

  > Chen S, Zhou Y, Chen Y, Gu J. fastp: an ultra-fast all-in-one FASTQ preprocessor. Bioinformatics. Sep 1 2018;34(17):i884-i890. doi: 10.1093/bioinformatics/bty560

- [Fastp latest](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC10989850/)

  > Chen S. Ultrafast one-pass FASTQ data preprocessing, quality control, and deduplication using fastp. Imeta. May 8 2023;2(2):e107. doi: 10.1002/imt2.107

- [FLASH](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3198573/)

  > Magoč T, Salzberg SL. FLASH: fast length adjustment of short reads to improve genome assemblies. Bioinformatics. Nov 1 2011;27(21):2957-63. doi:10.1093/bioinformatics/btr507

- [GTDB-Tk](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC9710552/)

  > Chaumeil PA, Mussig AJ, Hugenholtz P, Parks DH. GTDB-Tk v2: memory friendly classification with the genome taxonomy database. Bioinformatics. Nov 30 2022;38(23):5315-5316. doi:10.1093/bioinformatics/btac672

- [Hostile](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC10749771/)

  > Constantinides B, Hunt M, Crook DW. Hostile: accurate decontamination of microbial host sequences. Dec 1 2023;39(12):btad728. doi: 10.1093/bioinformatics/btad728

- [Kraken](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4053813/)
  > Wood DE, Salzberg SL. Kraken: ultrafast metagenomic sequence classification using exact alignments. Genome Biology. 2014/03/03 2014;15(3):R46. doi:10.1186/gb-2014-15-3-r46

- [Kraken 2](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1891-0)

  > Wood DE, Lu J, Langmead B. Improved metagenomic analysis with Kraken 2. Genome Biology. 2019/11/28 2019;20(1):257. doi:10.1186/s13059-019-1891-0

- [mlst](https://github.com/tseemann/mlst)

  > Seeman T. mlst. Github: <https://github.com/tseemann/mlst>

- [Pilon](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4237348/)

  > Walker BJ, Abeel T, Shea T, et al. Pilon: An Integrated Tool for Comprehensive Microbial Variant Detection and Genome Assembly Improvement. PLOS ONE. 2014;9(11):e112963. doi:10.1371/journal.pone.0112963

- [PubMLST](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6192448/)

  > Jolley KA, Bray JE, Maiden MCJ. Open-access bacterial population genomics: BIGSdb software, the PubMLST.org website and their applications. Wellcome Open Res. 2018;3:124. doi:10.12688/wellcomeopenres.14826.1

- [Prokka](https://academic.oup.com/bioinformatics/article/30/14/2068/2390517)

  > Seemann T. Prokka: rapid prokaryotic genome annotation. Bioinformatics. Jul 15 2014;30(14):2068-9. doi:10.1093/bioinformatics/btu153

- [QUAST](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3624806/)

  > Gurevich A, Saveliev V, Vyahhi N, Tesler G. QUAST: quality assessment tool for genome assemblies. Bioinformatics. Apr 15 2013;29(8):1072-5. doi:10.1093/bioinformatics/btt086

- [QUAST latest](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6022658/)

  > Mikheenko A, Prjibelski A, Saveliev V, Antipov D, Gurevich A. Versatile genome assembly evaluation with QUAST-LG. Jul 1 2018;34(13):i142-i150. doi: 10.1093/bioinformatics/bty266

- [RDP Classifier](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC11008197/)

  > Wang Q, Cole JR. Updated RDP taxonomy and RDP Classifier for more accurate taxonomic classification. Microbiol Resour Announc. Apr 11 2024;13(4):e0106323. doi: 10.1128/mra.01063-23

- [RNAmmer](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1888812/)

  > Lagesen K, Hallin P, Rødland EA, Staerfeldt HH, Rognes T, Ussery DW. RNAmmer: consistent and rapid annotation of ribosomal RNA genes. Nucleic Acids Res. 2007;35(9):3100-8. doi:10.1093/nar/gkm160

- [SAMtools](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2723002/)

  > Li H, Handsaker B, Wysoker A, et al. The Sequence Alignment/Map format and SAMtools. Bioinformatics. Aug 15 2009;25(16):2078-9. doi:10.1093/bioinformatics/btp352

- [SeqFu](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8148589/)

  > Telatin A, Fariselli P, Birolo G. SeqFu: A Suite of Utilities for the Robust and Reproducible Manipulation of Sequence Files. May 7 2021;8(5):59. doi: 10.3390/bioengineering8050059

- [SeqKit](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5051824/)

  > Shen W, Le S, Li Y, Hu F. SeqKit: A Cross-Platform and Ultrafast Toolkit for FASTA/Q File Manipulation. Oct 5 2016;11(10):e0163962. doi: 10.1371/journal.pone.0163962

- [SeqKit latest](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC11183193/)

  > Shen W, Sipos B, Zhao L. SeqKit2: A Swiss army knife for sequence and alignment processing. Imeta. Apr 5 2024;3(3):e191. doi: 10.1002/imt2.191

- [SKESA](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-018-1540-z)

  > Souvorov A, Agarwala R, Lipman DJ. SKESA: strategic k-mer extension for scrupulous assemblies. Genome Biology. 2018/10/04 2018;19(1):153. doi:10.1186/s13059-018-1540-z

- [SPAdes](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3342519/)

  > Bankevich A, Nurk S, Antipov D, et al. SPAdes: a new genome assembly algorithm and its applications to single-cell sequencing. J Comput Biol. May 2012;19(5):455-77. doi:10.1089/cmb.2012.0021

- [SPAdes latest](https://pubmed.ncbi.nlm.nih.gov/32559359/)
  > Prjibelski A, Antipov D, Meleshko D, Lapidus A, Korobeynikov A. Using SPAdes De Novo Assembler. Curr Protoc Bioinformatics. Jun 2020;70(1):e102. doi: 10.1002/cpbi.102

- [Trimmomatic](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4103590/)

  > Bolger AM, Lohse M, Usadel B. Trimmomatic: a flexible trimmer for Illumina sequence data. Bioinformatics. Aug 1 2014;30(15):2114-20. doi:10.1093/bioinformatics/btu170

## Software packaging/containerisation tools

- [Anaconda](https://anaconda.com)

  > Anaconda Software Distribution. Computer software. Vers. 2-2.4.0. Anaconda, Nov. 2016. Web.

- [Bioconda](https://pubmed.ncbi.nlm.nih.gov/29967506/)

  > Grüning B, Dale R, Sjödin A, Chapman BA, Rowe J, Tomkins-Tinch CH, Valieris R, Köster J; Bioconda Team. Bioconda: sustainable and comprehensive software distribution for the life sciences. Nat Methods. 2018 Jul;15(7):475-476. doi: 10.1038/s41592-018-0046-7. PubMed PMID: 29967506.

- [BioContainers](https://pubmed.ncbi.nlm.nih.gov/28379341/)

  > da Veiga Leprevost F, Grüning B, Aflitos SA, Röst HL, Uszkoreit J, Barsnes H, Vaudel M, Moreno P, Gatto L, Weber J, Bai M, Jimenez RC, Sachsenberg T, Pfeuffer J, Alvarez RV, Griss J, Nesvizhskii AI, Perez-Riverol Y. BioContainers: an open-source and community-driven framework for software standardization. Bioinformatics. 2017 Aug 15;33(16):2580-2582. doi: 10.1093/bioinformatics/btx192. PubMed PMID: 28379341; PubMed Central PMCID: PMC5870671.

- [Docker](https://dl.acm.org/doi/10.5555/2600239.2600241)

- [Singularity](https://pubmed.ncbi.nlm.nih.gov/28494014/)
  > Kurtzer GM, Sochat V, Bauer MW. Singularity: Scientific containers for mobility of compute. PLoS One. 2017 May 11;12(5):e0177459. doi: 10.1371/journal.pone.0177459. eCollection 2017. PubMed PMID: 28494014; PubMed Central PMCID: PMC5426675.

### Citation Information

Citations were created with the help of EndNote and were exported in AMA format.
