<h1>
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="images/wf-paired-end-illumina-assembly_logo_dark.png">
    <img alt="bacterial-genomics/wf-paired-end-illumina-assembly" src="images/wf-paired-end-illumina-assembly_logo_light.png">
  </picture>
</h1>

![workflow](images/wf-paired-end-illumina-assembly_workflow.png)

> _General schematic of the steps in the workflow_

## Requirements

- [`Nextflow`](https://www.nextflow.io/docs/latest/getstarted.html#installation) (`>=22.04.3`)
- [`Docker`](https://docs.docker.com/engine/installation/) or [`Singularity`](https://www.sylabs.io/guides/3.0/user-guide/) (`>=3.8.0`)

## Install on our HPC

```bash
git clone https://github.com/bacterial-genomics/wf-paired-end-illumina-assembly.git $LAB_HOME/workflows
```

## Setup Singularity environment variables - For Aspen Cluster

```bash
# Add to $HOME/.bashrc
SINGULARITY_BASE=/scicomp/scratch/$USER

export SINGULARITY_TMPDIR=$SINGULARITY_BASE/singularity.tmp

export SINGULARITY_CACHEDIR=$SINGULARITY_BASE/singularity.cache

export NXF_SINGULARITY_CACHEDIR=$SINGULARITY_BASE/singularity.cache

mkdir -pv $SINGULARITY_TMPDIR $SINGULARITY_CACHEDIR
```

Reload .bashrc

```bash
source ~/.bashrc
```

# Run Workflow

Before running workflow on new data, the workflow should be run on the built-in test data to make sure everything is working properly. It will also download all dependencies to make subsequent runs much faster.

```bash
cd $LAB_HOME/workflows/wf-paired-end-illumina-assembly

module load nextflow

nextflow run main.nf -profile singularity,test --outdir results
```

To minimize typing all the parameters above, a bash script was created for UGE HPCs. It can take PE FastQ files from selected directory OR if FastQ files not found in that directory, it will look in subdirectories for FastQ files. To run:

```bash
# Uses SPAdes assembler
run_assembly.uge-nextflow INPUT_DIRECTORY
```

Example analysis using Nextflow command:

```bash
nextflow run main.nf \
  -profile singularity \
  --input INPUT_DIRECTORY \
  --outdir OUTPUT_DIRECTORY \
  --assembler <spades|skesa>
```

## Help menu of all options

```bash
nextflow run main.nf --help
```

Test data was generated by taking top 1 million lines (=250k reads) of SRA data SRR16343585. (Note: This requires SRA Toolkit)

```bash
fasterq-dump SRR16343585
head -1000000 SRR16343585_1.fastq > test_R1.fastq
head -1000000 SRR16343585_2.fastq > test_R2.fastq
pigz test_R{1,2}.fastq
```
