<h1>
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="images/wf-paired-end-illumina-assembly_logo_dark.png">
    <img alt="bacterial-genomics/wf-paired-end-illumina-assembly" src="images/wf-paired-end-illumina-assembly_logo_light.png">
  </picture>
</h1>

![workflow](images/wf-paired-end-illumina-assembly_workflow.png)

> _General schematic of the steps in the workflow_

## Requirements

- [`Nextflow`](https://www.nextflow.io/docs/latest/getstarted.html#installation) (`>=22.04.3`)
- [`Docker`](https://docs.docker.com/engine/installation/) or [`Singularity`](https://www.sylabs.io/guides/3.0/user-guide/) (`>=3.8.0`)

## Install workflow locally

```bash
git clone https://github.com/bacterial-genomics/wf-paired-end-illumina-assembly.git
```

# Run workflow

Before running workflow on new data, the workflow should be run on the built-in test data to make sure everything is working properly. It will also download all dependencies to make subsequent runs much faster.

```bash
cd wf-paired-end-illumina-assembly/

nextflow run main.nf -profile docker,test --outdir results
```

## Usage

```bash
nextflow run main.nf \
  -profile <docker|singularity> \
  --input INPUT_DIRECTORY \
  --outdir OUTPUT_DIRECTORY \
  --assembler <spades|skesa>
```

## Local resource management

> [!WARNING]
> Lowering the number of CPUs to use in the workflow increases the run time, but it will still successfully complete. If RAM (memory) is decreased too low, however, the task will fail. This workflow works for samples sequenced to a reasonable depth of coverage with 6.GB RAM and 1 CPU.
>
> If you do run out of RAM, the host removal and downsampling steps are likely to be the issue(s). To turn each off, you can append `--depth 0` to turn off downsampling, and `--host_remove skip` to turn off host removal.

Default settings on available number of CPUs and size of RAM (memory) might be inappropriate for your local system, but you can change both. When running locally on a desktop or laptop, `--max_cpus` and `--max_memory` can be specified, which is often to either maximize runtime speed or to ensure the workflow does not take over your whole system resources and make using it in other ways difficult. Below, CPUs is set to 4 and RAM is set to 16.GB (for 16 GB). The formatting requirement for max_memory might seem unusual but must be followed (e.g., 16.GB, 64.GB).

```bash
nextflow run main.nf \
  -profile docker \
  --input INPUT_DIRECTORY \
  --outdir OUTPUT_DIRECTORY \
  --max_cpus 4 \
  --max_memory 16.GB
```

> [!TIP]
> Web browsers (e.g., Apple Safari, Google Chrome, Microsoft Edge, Mozilla Firefox, Opera) often consume quite a lot of RAM (memory). If you avoid having those open while running the workflow, you can allocate much more `--max_memory`.

## Update local workflow to the latest stable version

```bash
cd wf-paired-end-illumina-assembly/
git pull
```

## Change local workflow to a [specific version](https://github.com/bacterial-genomics/wf-paired-end-illumina-assembly/tags) (e.g., v2.3.0)

```bash
cd wf-paired-end-illumina-assembly/
git checkout tags/v2.3.0
```

### Help menu of all workflow options

```bash
nextflow run main.nf --help --show_hidden_params
```

Test data was generated by taking top 1 million lines (=250k reads) of SRA data SRR16343585. (Note: This requires SRA Toolkit)

```bash
fasterq-dump SRR16343585
head -1000000 SRR16343585_1.fastq > test_R1.fastq
head -1000000 SRR16343585_2.fastq > test_R2.fastq
pigz test_R{1,2}.fastq
```
